src/ai/backend/manager/models/etcd.py

```diff
@@ -0,0 +1,257 @@
+from __future__ import annotations
+
+import logging
+from typing import TYPE_CHECKING, Mapping, Sequence
+
+import graphene
+import trafaret as t
+
+from ai.backend.common.logging import BraceStyleAdapter
+
+from ..api.exceptions import InvalidAPIParameters, ObjectNotFound
+from ..config import container_registry_iv
+from . import UserRole
+from .base import privileged_mutation
+
+if TYPE_CHECKING:
+    from .gql import GraphQueryContext
+
+log = BraceStyleAdapter(
+    logging.getLogger("ai.backend.manager.models.etcd")
+)  # type: ignore[name-defined]
+
+__all__: Sequence[str] = (
+    "ContainerRegistry",
+    "ContainerRegistries",
+    "CreateContainerRegistry",
+    "ModifyContainerRegistry",
+    "DeleteContainerRegistry",
+)
+
+ETCD_CONTAINER_REGISTRY_KEY = "config/docker/registry"
+
+
+class CreateContainerRegistryInput(graphene.InputObjectType):
+    url = graphene.String(required=True)
+    container_type = graphene.String(required=True)
+    project = graphene.String()
+    username = graphene.String()
+    password = graphene.String()
+
+
+class ModifyContainerRegistryInput(graphene.InputObjectType):
+    url = graphene.String(required=True)
+    container_type = graphene.String(required=True)
+    project = graphene.String()
+    username = graphene.String()
+    password = graphene.String()
+
+
+class ContainerRegistryConfig(graphene.ObjectType):
+    url = graphene.String(required=True)
+    container_type = graphene.String(required=True)
+    project = graphene.List(lambda: graphene.String)
+    username = graphene.String()
+    password = graphene.String()
+
+
+class ContainerRegistry(graphene.ObjectType):
+    hostname = graphene.String()
+    config = graphene.Field(ContainerRegistryConfig)
+
+    @classmethod
+    def from_row(cls, hostname: str, config: Mapping[str, str | list | None]) -> ContainerRegistry:
+        return cls(
+            hostname=hostname,
+            config=ContainerRegistryConfig(
+                url=config.get(""),
+                container_type=config.get("type"),
+                project=config.get("project", None),
+                username=config.get("username", None),
+                password=config.get("password", None),
+            ),
+        )
+
+    @classmethod
+    async def load_all(
+        cls,
+        ctx: GraphQueryContext,
+    ) -> Sequence[ContainerRegistry]:
+        log.info(
+            "ETCD.LOAD_ALL_CONTAINER_REGISTRIES (ak:{}, key:{})",
+            ctx.access_key,
+            ETCD_CONTAINER_REGISTRY_KEY,
+        )
+        raw_registries = await ctx.shared_config.etcd.get_prefix_dict(ETCD_CONTAINER_REGISTRY_KEY)
+        registries = t.Mapping(t.String, container_registry_iv).check(raw_registries)
```

Comment:
Seems this code is used as the same in other functions.
How about implement a method like `get_container_registry()` in SharedConfig?

Reply:
I agree with @fregataa. Let's make the CUD methods reusable across other parts of our codebase by moving them to `SharedConfig`.